### PR TITLE
rtt_rtc: allow to use RTC emulation together with RTT alarms

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

Some RTT implementations provide a separate overflow callback.
`rtt_rtc` was originally designed to work with the most basic RTT: It only needs a monotonic counter and an alarm callback that is shared between overflow and RTC alarm.

If we use the overflow callback, this frees up the alarm callback for direct use with the RTT API.

This means that if an RTT provides the alarm callback and the user treads it as a monotonic counter (no calls to `rtt_set_counter()`), both `rtc_rtt` and the `periph_rtt` API can be used together.

The user also must make sure to either use RTC or RTT alarms as they will overwrite each other.

But this allows to track & adjust the time without having to re-calculate RTT alarms if the time is changed due to synchronization.

### Testing procedure

So far I only tested `tests/periph_rtc` with `rtt_rtc` instead of `periph_rtc` both with and without the overflow callback support.
Both worked as expected.

### CODE REVIEW

**Comment by maribu** [comment]
Looks quite good to me. But let's wait for the feature freeze to end first.

**Comment by benpicco** [comment]
It's feature thaw again 😄

**Comment by maribu** [comment]
lgtm. let me do some testing to be sure :-)

**Comment by maribu** [comment]
Maybe it makes sense to reorder the commits, so that the feature gets introduced first and `rtt_rtc` is updated to use it afterwards?

**Comment by maribu** [comment]
There is something wrong on AVR:

```
2021-03-11 09:36:23,527 # START
2021-03-11 09:36:23,528 # main(): This is RIOT! (Version: 2021.04-devel-1026-g9991b-rtt_rtc)
2021-03-11 09:36:23,529 # 
2021-03-11 09:36:23,529 # RIOT RTC low-level driver test
2021-03-11 09:36:23,530 # This test will display 'Alarm!' every 2 seconds for 4 times
2021-03-11 09:36:23,531 #   Setting clock to   2020-02-28 23:59:57
2021-03-11 09:36:23,531 # Clock value is now   2022-01-01 00:00:-3
2021-03-11 09:36:23,532 #   Setting alarm to   2022-01-01 00:00:-1
2021-03-11 09:36:23,532 #    Alarm is set to   2025-01-01 00:00:-1
2021-03-11 09:36:23,533 #   Alarm cleared at   2022-01-01 00:00:-3
2021-03-11 09:36:25,488 #        No alarm at   2022-01-01 00:00:-1
2021-03-11 09:36:25,531 #   Setting alarm to   2022-01-01 00:00:01
2021-03-11 09:36:25,532 # 
```

Tested with:

```diff
diff --git a/tests/periph_rtc/Makefile b/tests/periph_rtc/Makefile
index e549720ff4..ebfc5bcc41 100644
--- a/tests/periph_rtc/Makefile
+++ b/tests/periph_rtc/Makefile
@@ -1,6 +1,6 @@
 include ../Makefile.tests_common
 
-FEATURES_REQUIRED = periph_rtc
+USEMODULE += rtt_rtc
 
 DISABLE_MODULE += periph_init_rtc
```

**Comment by benpicco** [comment]
This bug is also present on master :(

**Comment by benpicco** [comment]
The issue on AVR should be unrelated to this PR though.

**Comment by maribu** [comment]
Tested using patch from https://github.com/RIOT-OS/RIOT/pull/15763#issuecomment-796567283 on `wemos-zero` and `nucleo-f767zi` (both having `periph_rtt_overflow` and it still worked.

Tested again with

```patch
diff --git a/tests/periph_rtc/Makefile b/tests/periph_rtc/Makefile
index e549720ff4..aa1918b691 100644
--- a/tests/periph_rtc/Makefile
+++ b/tests/periph_rtc/Makefile
@@ -1,8 +1,9 @@
 include ../Makefile.tests_common
 
-FEATURES_REQUIRED = periph_rtc
+USEMODULE += rtt_rtc
 
 DISABLE_MODULE += periph_init_rtc
+FEATURES_BLACKLIST += periph_rtt_overflow
 
 USEMODULE += xtimer
 
```

Fails with this `assert()` blowing up:

https://github.com/RIOT-OS/RIOT/blob/2a3fbf38f60e27ec84c83e093aecbc38c879821c/cpu/stm32/periph/rtt_all.c#L175-L186

With this patch I get:

```patch
diff --git a/cpu/stm32/periph/rtt_all.c b/cpu/stm32/periph/rtt_all.c
index 9e4ff785ab..d9f341688b 100644
--- a/cpu/stm32/periph/rtt_all.c
+++ b/cpu/stm32/periph/rtt_all.c
@@ -26,6 +26,9 @@
 #include "periph/rtt.h"
 #include "stmclk.h"
 
+#define ENABLE_DEBUG 1
+#include "debug.h"
+
 /* this driver is only valid for STM CPUs that provide LPTIMERs */
 #if defined(LPTIM1)
 
@@ -174,6 +177,7 @@ void rtt_clear_overflow_cb(void)
 
 void rtt_set_alarm(uint32_t alarm, rtt_cb_t cb, void *arg)
 {
+    DEBUG("cb = %p, alarm = %u, RTT_MAX_VALUE = %u\n", (void *)cb, (unsigned)alarm, (unsigned)RTT_MAX_VALUE);
     assert(cb && !(alarm & ~RTT_MAX_VALUE));
 
     unsigned is = irq_disable();
```

```
READY
s
START
main(): This is RIOT! (Version: 2021.04-devel-1026-g9991b-ben)

RIOT RTC low-level driver test
This test will display 'Alarm!' every 2 seconds for 4 times
cb = 0x8001611, alarm = 32767, RTT_MAX_VALUE = 65535
  Setting clock to   2020-02-28 23:59:57
cb = 0x8001611, alarm = 74122, RTT_MAX_VALUE = 65535
0x8000fa5
*** RIOT kernel panic:
FAILED ASSERTION.

*** halted.
```

File: drivers/rtt_rtc/rtt_rtc.c [hunk 1]
```diff
@@ -65,7 +65,18 @@ static inline uint32_t _rtc_now(uint32_t now)
 
 static inline void _set_alarm(uint32_t now, uint32_t next_alarm)
 {
-    rtt_set_alarm(now + next_alarm, _rtt_alarm, NULL);
+    if (IS_USED(MODULE_PERIPH_RTT_OVERFLOW)) {
+        rtt_set_alarm(now + next_alarm, alarm_cb, alarm_cb_arg);
+    } else {
+        rtt_set_alarm(now + next_alarm, _rtt_cb, NULL);
```
Comment (maribu) [inline drivers/rtt_rtc/rtt_rtc.c:71]: This line is the offender.

GDB:
```
(gdb) p next_alarm
$1 = 32768
(gdb) p now
$2 = <optimized out>
```

Console:
```
2021-04-20 11:30:26,644 # cb = 0x8001611, alarm = 65536, RTT_MAX_VALUE = 65535
```

**Comment by benpicco** [comment]
Thank you for tracking this down!

**Comment by MrKevinWeiss** [comment]
[Basic PR](https://ci.riot-os.org/hil/blue/organizations/jenkins/RIOT_tests/detail/RIOT_tests/12/pipeline/)
[With patch](https://ci.riot-os.org/hil/blue/organizations/jenkins/RIOT_tests/detail/RIOT_tests/11/pipeline/)

```diff
diff --git a/tests/periph_rtc/Makefile b/tests/periph_rtc/Makefile
index e549720ff4..aa1918b691 100644
--- a/tests/periph_rtc/Makefile
+++ b/tests/periph_rtc/Makefile
@@ -1,8 +1,9 @@
 include ../Makefile.tests_common
 
-FEATURES_REQUIRED = periph_rtc
+USEMODULE += rtt_rtc
 
 DISABLE_MODULE += periph_init_rtc
+FEATURES_BLACKLIST += periph_rtt_overflow
 
 USEMODULE += xtimer

```

**Comment by maribu** [comment]
@MrKevinWeiss My conclusion is that all tests passing in `master` still pass with this PR, except for when there are issues with flashing / connecting / `No such saved stash ...` issue in the tool. Is this correct?

For me, it now works on my Nucleo-F767ZI. So, if there are no objections, I'd say this can go in.

**Comment by maribu** [comment]
Btw: Feel free to squash.

**Comment by MrKevinWeiss** [comment]
> except for when there are issues with flashing / connecting / `No such saved stash ...` issue in the tool. Is this correct?

No saved stash is a build error. The flashing issues should have been resolved, at least nighties seem to survive pretty well on all boards.

Does it also work with the patch on the nucleo-f767zi?

**Comment by MrKevinWeiss** [comment]
Either way I am not going to be blocking with the current confidence in the HiL setup.

**Comment by benpicco** [comment]
The issues @MrKevinWeiss uncovered are real though, I also see this fail on my `nucleo-l031k6`.

**Comment by fjmolinas** [comment]
> The issues @MrKevinWeiss uncovered are real though, I also see this fail on my `nucleo-l031k6`.

@benpicco I don't quite understand the issues you point out, on what hardware can I reproduce?

>If we use the overflow callback, this frees up the alarm callback for direct use with the RTT API.

>This means that if an RTT provides the alarm callback and the user treads it as a monotonic counter (no calls to rtt_set_counter()), both rtc_rtt and the periph_rtt API can be used together.

>The user also must make sure to either use RTC or RTT alarms as they will overwrite each other.

Can all of this be documented somewhere?

**Comment by fjmolinas** [comment]
This one now needs a rebase @benpicco!

**Pull request closed without merging**